main/src/library/List.flix

```diff
@@ -0,0 +1,527 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace List {
+
+    ///
+    /// Returns true iff `xs` is the empty list, i.e. `Nil`.
+    ///
+    def null[a](xs: List[a]): Bool = match xs with {
+        case Nil => true
+        case _ => false
+    }
+
+    ///
+    /// Returns the first element of `xs`.
+    /// If `xs` is empty the semantics are undefined.
+    ///
+    def head[a](xs: List[a]): a = match xs with {
+        case x :: _ => x
+    }
+
+    ///
+    /// Returns the last element of `xs`.
+    /// If `xs` is empty the semantics are undefined.
+    ///
+    def last[a](xs: List[a]): a = match xs with {
+        case x :: Nil => x
+        case x :: rs => last(rs)
+    }
+
+    ///
+    /// Returns the list consisting of every element of `xs` except the last element.
+    /// If `xs` is empty the semantics are undefined.
+    ///
+    def init[a](xs: List[a]): List[a] = match xs with {
+        case x :: Nil => Nil
+        case x :: rs => x :: init(rs)
+    }
+
+    ///
+    /// Returns the list consisting of every element of `xs` except the first element.
+    /// If `xs` is empty the semantics are undefined.
+    ///
+    def tail[a](xs: List[a]): List[a] = match xs with {
+        case _ :: rs => rs
+    }
+
+    ///
+    /// Returns the first `n` elements of `xs` in a list.
+    /// If `n < 0` or `n > length(xs)` the semantics are undefined.
+    ///
+    def prefix[a](n: Int32, xs: List[a]): List[a] = match (n, xs) with {
+        case (0, _) => Nil
+        case (i, x :: rs) => x :: prefix(i-1, rs)
+    }
+
+    ///
+    /// Returns the last `n` elements of `xs` in a list.
+    /// If `n < 0` or `n > length(xs)` the semantics are undefined.
+    ///
+    def suffix[a](n: Int32, xs: List[a]): List[a] = match (length(xs) - n, xs) with {
+        case (0, _) => xs
+        case (_, x :: rs) => suffix(n, rs)
+    }
+
+    ///
+    /// Returns the length of `xs`.
+    ///
+    def length[a](xs: List[a]): Int32 = match xs with {
+        case Nil => 0
+        case x :: rs => 1 + length(rs)
+    }
+
+    ///
+    /// Returns `ys` appended to `xs`.
+    ///
+    def append[a](xs: List[a], ys: List[a]): List[a] = match xs with {
+        case Nil => ys
+        case x :: rs => x :: append(rs, ys)
+    }
+
+    ///
+    /// Returns the element at position `p` in `xs`.
+    /// If `p` < 0 or `p` > length(xs)-1 the semantics are undefined.
+    ///
+    def at[a](p: Int32, xs: List[a]): a = match (p, xs) with {
```

Comment:
There doesn't seem to be a good naming convention for this. Scala uses `apply` (i.e. parenthesis indexing), Haskell uses `!!`, OCaml uses `nth`, Racket doesn't seem to have it. C++ has `at` for vectors.